driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,529 @@
+         * {@link ErrorAwarePolicyTracker} for more information.
+         *
+         * @param errorTracker the tracker to plug in.
+         * @return this {@link Builder} instance, for method chaining.
+         */
+        public Builder withErrorTracker(ErrorAwarePolicyTracker errorTracker) {
+            this.errorTracker = errorTracker;
+            return this;
+        }
+
+        /**
+         * Defines the time during which a host is excluded by the tracker.
+         * <p/>
+         * When a host's error rate overpasses the threshold specified in {@link #withInclusionThreshold}, the host
+         * will be considered excluded, and this for a certain period of time defined by this method. After this period,
+         * a previously excluded host will be included again until it reaches the threshold again.
+         * <p/>
+         * If the error tracker, during an update, detects that a currently excluded host continues to overpass the
+         * error rate threshold, this host's exclusion time will be reset to the start.
+         * <p/>
+         * Default value for the retry period is 5 seconds.
+         *
+         * @param retryPeriod the period to use for refreshing the ErrorTracker.
+         * @return this {@link Builder} instance, for method chaining.
+         */
+        public Builder withRetryPeriod(int retryPeriod) {
+            this.retryPeriod = retryPeriod;
+            return this;
+        }
+
+        /**
+         * Creates the {@link ErrorAwarePolicy} instance.
+         * @return the newly created {@link ErrorAwarePolicy}.
+         */
+        public ErrorAwarePolicy build() {
+            return new ErrorAwarePolicy(this);
+        }
+    }
+
+    /**
+     * Default implementation for the {@link ErrorAwarePolicyTracker}.
+     * <p/>
+     * This tracker will take note of the errors happening for each hosts, and will accordingly decide
+     * whether a hosts has exceeded a configurable threshold of errors per seconds. If the number of
+     * errors the host had was greater than or equal to the threshold, the host will be excluded.
+     */
+    public static class PerHostErrorTracker implements ErrorAwarePolicyTracker {
+
+        private ConcurrentMap<Host, AtomicInteger> errors = new ConcurrentHashMap<Host, AtomicInteger>();
+
+        private Map<Host, ErrorRecord> errorsPerSec = new HashMap<Host, ErrorRecord>();
+
+        private long retryPeriod;
+
+        private int errorsPerSecThreshold;
+
+        private double errorsThresholdRate;
+
+        private long updatePeriod;
+
+        @Override
+        public void notifyError(Host host, Statement statement, Exception exception) {
+            // should just keep counts of errors for each hosts
+            if (!shouldAccountForError(host, statement, exception)) {
+                return;
+            }
+
+            AtomicInteger ai = errors.putIfAbsent(host, new AtomicInteger(1));
```

Comment:
This allocates a new `AtomicInteger` instance on each call, it will only be really used the first time.